Shaders : Parameter plugs no longer dynamic

danieldresser-ie opened the pull request:
This is experimental enough that I'm just putting code here for reference, and I'll discuss with John elsewhere.  It's something we do want to do, but I'm not sure if it will be now, or if it will look like this.

johnhaddon commented:
>    script.addChild( Shader( nodeName, shaderName ) )

My gut reaction to this was negative, but I'm not sure why. The only rational reason I can come up with is that keeping the `loadShader()` call in the serialisation provides us with some ability to monkey-patch things in a compatibility config later. It feels like that might be a good thing to keep, at least for a little while...

I guess the other thing is that I saw this PR as being about just doing what we needed to optimise the serialisation format, rather than changing the API.

> As far as I know, we're now saying that shader nodes must load a shader during construction, and this change would make that explicit.

Actually, I don't think this is true. I think it might still be useful to construct a shader and then add plugs yourself (certainly for the base class), and also to construct a shader and then call `loadShader()` at a later date. If you really want to enforce loading at construction, then `loadShader()` should probably be renamed to `reload()` or something. But I like the current flexibility, and I don't see much tangible benefit to changing things...

> The way we serialize lights is currently pretty ugly? It looks like there is no serialized loadShader call at all, so if you the renderer modifies the parameters available on a light, you must manually recreate your light in order to pick this up?

Yeah, lights themselves are weird anyway, because there's an awkward overlap in functionality between them and shaders. Shader nodes know how to load a shader and turn a network into attributes to be assigned, and Light nodes have [half-assed](https://github.com/GafferHQ/gaffer/blob/master/src/GafferArnold/ArnoldLight.cpp#L111) versions of that, along with the ability to create a location and assign the attributes.

Really, I think lights should just be about the latter, and should delegate to shaders to do the loading and network generation. That's exactly how I implemented the [OSLLight](https://github.com/GafferHQ/gaffer/blob/master/src/GafferArnold/ArnoldLight.cpp#L111) node recently, and I'd love to try pushing that model up onto the Light base class and implementing the other lights the same way.

It's not something to do right now, but I thought it'd be useful to let you know what I'm thinking...

> It feels like it would be more elegant to handle this the same way as Shader nodes, either by serializing a loadShader, or if you like 1), adding the same constructor argument.

Agreed.
 
> 2) might be a bit hard to make backwards compatible, since unlike Shader nodes, we don't have the serialized loadShader in old files to trigger disabling the Dynamic flag on all existing parameters. Disabling Dynamic during serialization of lights seems like it could work - then the first time you load an old script, it loads the old, it gets saved the new way, and the second time you open it with a new Gaffer, it gets loaded the new way ( and actually resynchronizes to the renders definition of the light type ).

Sounds like a plan to me.

> Not sure how to weigh up the goals of "minimal PR" vs "now is the time to break things if we ever want to clean this up". I'm game to try changing lights as well if you think it's a good idea.

I would vote for sorting out the Light serialisation as you suggest, but keeping the explicit `loadShader()` calls following the constructor for now. I'm not sure why, but I'm not quite ready to let go of those yet...

danieldresser-ie commented:
This ended up being a bunch of little changes to fix all the lights as well - I got pulled off onto some production stuff in the middle of working on this, so I'm a bit worried that I might have forgotten something that needed doing, but the basic framework of this seems to be working now.

Definitely needs a careful review.

johnhaddon commented on src/GafferAppleseedModule/GafferAppleseedModule.cpp:
```diff
@@ -50,12 +50,49 @@
+			}
+		}
+
+		// \todo - Accessing this plug by explicit name is ugly, but the accessor is private.
```
It seems a pity that we're having to make serialisers for each light subclass individually, rather than just having one LightSerialiser that works for all of them. Is this the only reason for that, or are there others too? In the interests of moving things closer to where we want them to be, I'd be happy to see this hack included in a single central LightSerialiser class if we can make that work.

I think another issue would be that the Light base class doesn't have a `loadShader()` method, but we can look for it is a python attribute in the same way as the `ValuePlugSerialiser` [looks for `getValue()`](https://github.com/GafferHQ/gaffer/blob/master/src/GafferBindings/ValuePlugBinding.cpp#L141).

danieldresser-ie replied:
This would also require standardizing on a way to get the shader name.  Currently we have:

public: OSLLight::shaderNamePlug() 
private: ArnoldLight::shaderNamePlug() 
private: AppleseedLight::modelPlug()

If you want to include this in the base class, should we just standardize on lights having a loadShader and shaderNamePlug(), and make these virtual in the base class?  ( I kinda feel like shaderName could just be a member variable instead of a plug, but you probably want to keep the possibility of driving it dynamically somehow, since you didn't like the idea of passing it to the constructor. )

johnhaddon replied:
For now I was thinking we could just always access the plug by the `getChild<StringPlug>( name )` method you used here, looking for the three names in question. That'll be ugly for sure, but it gives us a single serialiser, and a single place to put a big comment saying that we should standardise.

I think the consideration of plug vs member should be done when refactoring the Light base class to adopt the OSLLight's "defer to internal Shader node" approach across the board. I think that's out of the scope of this PR though - we've already got the serialisation to where we want it to be, the rest can wait...

johnhaddon commented:
One more thing : could you cherry-pick over Andrew's test cases from #2454 to demonstrate that they are indeed fixed by this PR?

danieldresser-ie commented:
OK, I think I've addressed all comments.  SplinePlug serialisation seems to be fine, both in an automated test and in inspecting the results I get from a copy/paste.

Not sure what you're thinking of in terms of adversarial testing. I would usually try it on some production cases, but in this case, most of the complexity is in the light serialisation stuff, which we don't use at IE, so that doesn't test much.  I'll try to get Andrew to take a look tomorrow.

danieldresser-ie commented:
I should probably mention that after doing everything I thought I needed to, and getting all tests passing, I did one more round of interactive testing, and realized there was still an issue with old out plugs.

I have now fixed this, but I suppose it makes it harder to argue that I have confidence that this is working.

johnhaddon commented:
> SplinePlug serialisation seems to be fine, both in an automated test and in inspecting the results I get from a copy/paste.

Great! Looks like I was worried for nothing. Seems like we always serialise a `postConstructor()` (where the SplinePlug hack is) even if a plug isn't Dynamic and we don't serialise a constructor.

> Not sure what you're thinking of in terms of adversarial testing. I would usually try it on some production cases, but in this case, most of the complexity is in the light serialisation stuff, which we don't use at IE, so that doesn't test much.

I think testing some production cases would be well worth it. As you say, some of the bigger changes are in the lights and won't affect IE, but we still want to be absolutely sure our lookdev will survive the changes. I'll do some manual testing with lights at this end...

johnhaddon commented:
Did a bit of testing on the lights side of things today, and didn't see any issues. Where are we at with testing at your end? Time to merge?

danieldresser-ie commented:
Yeah, I did some testing here, and aside from the unrelated serialization issue with internal software I noted in the email, everything looks good.

Andrew has now given this a skim, and didn't spot anything wrong, so I guess it's time to merge.

johnhaddon commented:
OK, merged! Thanks!
